In [327]:
import numpy as np
from numpy.linalg import inv
import math
import itertools
import pandas as pd

import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

time: 1.39 s


In [2]:
%install_ext https://raw.github.com/cpcloud/ipython-autotime/master/autotime.py
%load_ext autotime

/Users/Amiros/anaconda/lib/python3.5/site-packages/IPython/core/magics/extension.py:47: UserWarning: %install_ext` is deprecated, please distribute your extension as a python package.
  "as a python package.", UserWarning)


Installed autotime.py. To use it, type:
  %load_ext autotime


In [10]:
#load train data as numpy array
X_train = np.genfromtxt('/Users/Amiros/GitHub/Machine Learning for Data Science/hw3/data/gaussian_process/X_train.csv',delimiter=',')
y_train = np.genfromtxt('/Users/Amiros/GitHub/Machine Learning for Data Science/hw3/data/gaussian_process/y_train.csv',delimiter=',')

#load test data as numpy array
X_test = np.genfromtxt('/Users/Amiros/GitHub/Machine Learning for Data Science/hw3/data/gaussian_process/X_test.csv',delimiter=',')
y_test = np.genfromtxt('/Users/Amiros/GitHub/Machine Learning for Data Science/hw3/data/gaussian_process/y_test.csv',delimiter=',')

time: 114 ms


## 1-A

In [440]:
# kernel matrix restricted points in Training data
def K_restricted(X_train, b):
    k = np.linalg.norm(X_train[None,:,:]-X_train[:,None,:],axis=2)
    return np.exp(-1/b * k)

def G_process(X_train, y_train, X_test, b, sigma):
    K_n = K_restricted(X_train, b)
    
    I = np.identity(X_train.shape[0])
    c = np.linalg.inv((sigma) * I + K_n)
    
    m = []
    v = []
    for i in range(X_test.shape[0]):
    # Calculate mean
        distance = np.sqrt(np.sum((X_test[i] - X_train) ** 2, axis=1))
        G_Kernel = np.hstack(np.exp((-1/b) * distance).reshape(1,-1))
        w = np.dot(G_Kernel, c)
        a = np.dot(w, y_train)
        m.append(a)

    return m

time: 22.1 ms


## 1-B

In [441]:
b_list=[5,7,9,11,13,15]
sigma_list=[.1,.2,.3,.4,.5,.6,.7,.8,.9,1]
c = list(itertools.product(b_list, sigma_list))

y_predict = []
for i in range(len(c)):
    y_predict.append(G_process(X_train, y_train, X_test, c[i][0], c[i][1]))
y_predict = np.array(y_predict)

time: 1.95 s


In [442]:
def RMSE(y_predict, y_test, c):
    rmse = []
    
    for i in range(len(c)):
        a = math.sqrt(np.sum((y_predict[i] - y_test)**2) / 42)
        rmse.append(a)  
    return  rmse

time: 4.23 ms


In [443]:
rmse = RMSE(y_predict, y_test, c)
rmse_table = pd.DataFrame(
    {'parameters': c, 'rmse_value': rmse})
rmse_table

parameters  rmse_value
0    (5, 0.1)    1.938372
1    (5, 0.2)    1.933602
2    (5, 0.3)    1.930739
3    (5, 0.4)    1.930309
4    (5, 0.5)    1.931505
5    (5, 0.6)    1.933778
6    (5, 0.7)    1.936804
7    (5, 0.8)    1.940386
8    (5, 0.9)    1.944405
9      (5, 1)    1.948782
10   (7, 0.1)    1.940853
11   (7, 0.2)    1.934726
12   (7, 0.3)    1.933197
13   (7, 0.4)    1.934424
14   (7, 0.5)    1.937161
15   (7, 0.6)    1.940821
16   (7, 0.7)    1.945119
17   (7, 0.8)    1.949906
18   (7, 0.9)    1.955098
19     (7, 1)    1.960644
20   (9, 0.1)    1.941556
21   (9, 0.2)    1.935956
22   (9, 0.3)    1.936128
23   (9, 0.4)    1.938905
24   (9, 0.5)    1.942999
25   (9, 0.6)    1.947915
26   (9, 0.7)    1.953441
27   (9, 0.8)    1.959475
28   (9, 0.9)    1.965958
29     (9, 1)    1.972848
30  (11, 0.1)    1.941658
31  (11, 0.2)    1.937350
32  (11, 0.3)    1.939142
33  (11, 0.4)    1.943281
34  (11, 0.5)    1.948611
35  (11, 0.6)    1.954755
36  (11, 0.7)    1.961562
37  (11, 0.8)    1.968956
38  (11, 0.9)    1.976881
39    (11, 1)    1.985292
40  (13, 0.1)    1.941612
41  (13, 0.2)    1.938869
42  (13, 0.3)    1.942123
43  (13, 0.4)    1.947501
44  (13, 0.5)    1.954042
45  (13, 0.6)    1.961467
46  (13, 0.7)    1.969661
47  (13, 0.8)    1.978548
48  (13, 0.9)    1.988065
49    (13, 1)    1.998149
50  (15, 0.1)    1.941590
51  (15, 0.2)    1.940463
52  (15, 0.3)    1.945042
53  (15, 0.4)    1.951609
54  (15, 0.5)    1.959397
55  (15, 0.6)    1.968194
56  (15, 0.7)    1.977894
57  (15, 0.8)    1.988404
58  (15, 0.9)    1.999638
59    (15, 1)    2.011511

time: 53 ms


## 1-C

In [444]:
rmse_table.min()

parameters    (5, 0.1)
rmse_value     1.93031
dtype: object

time: 5.46 ms


The best solution is for b= 5 and sigma = 0.1 with rmse value of 1.93031.

This approach comapring to homework 1 gives lower rmse, therefore more accurate result.

## 1-D

In [445]:
# aa = X_train[:,3]
# bb = X_test[:,3]

time: 1.94 ms


# 2